### DISCUSSION TITLE
Better developer experience with FileCacheAdapter

### DISCUSSION THREAD

**Discussion by oliversalzburg** [mikro-orm/mikro-orm#2169] [Ideas]
While working on a metadata-cache related issue, I found that the default `FileCacheAdapter` is maybe not ideal in how it behaves.

At this time, if a cached item is mismatched with the current source file, the cached item is simply ignored and the metadata collection process happens instead. For a developer who has enabled the metadata cache, this is unexpected. It can lead to an apparently working system, even though the cache is not utilized at all.

Similarly, if the cached item points to an origin that no longer exists, the cached item will also not be used, as no hash can be calculated to compare against.

My proposal is to replace or offer a replacement for the `FileCacheAdapter`, which provides a better developer experience in cases where things aren't going as planned.

I am aware that I can just provide my own adapter and implement one that behaves exactly like I want, but I feel the default experience in MikroORM could be improved in this area, by providing a different adapter implementation itself.

From what I understand, the metadata cache is only generated through the CLI (and never from regular operations of the main application). So if I enable the metadata cache in my application and a file is missing from it, that seems like an actual issue to me that should be surfaced or explicitly ignored through configuration, as I've apparently forgotten to generate or re-generate the cache.

I could imagine providing a `StrictFileCacheAdapter` that's almost identical to current `FileCacheAdapter`, but `throw`s on validation failures. In general, the adapter should be enriched with (debug) logging to surface any failures or mismatches.

As a side note, looking at all of this, I was wondering if it would make sense to have a component that acts as a metadata provider, but reads from the metadata cache? My idea being, when I have the cache, I don't really want to load a different provider to read metadata, as I already have it.
I would use one config with the real metadata provider for the CLI, generate my cache and then use a different provider to only read the cache at runtime.

**Comment by B4nan** [comment]
> At this time, if a cached item is mismatched with the current source file, the cached item is simply ignored and the metadata collection process happens instead. For a developer who has enabled the metadata cache, this is unexpected. It can lead to an apparently working system, even though the cache is not utilized at all.

This is how cache invalidation works, for me this is a feature, not an issue, as it works very well. But maybe I just did not fully understand what you would expect. Sounds weird to talk about DX here, but I guess I missed your point.

A bit of context - the slowness in ts-morph reflection is in loading/analyzing the project, it is not about doing "more properties or more classes". Therefore it makes very little sense to handle differences in "some part of the cache is invalid" and "the whole cache is invalid".

Overall this sounds like you would like to use cache even when there are no origin entities - that itself sounds quite weird and error prone to me, rather than a good DX, but I might be wrong. 

> I am aware that I can just provide my own adapter and implement one that behaves exactly like I want, but I feel the default experience in MikroORM could be improved in this area, by providing a different adapter implementation itself.

If you see room for improvement, do your own implementation, then we can talk about adding it to the code base or even replacing the defaults. We can change some internals too if there would be a need for it.

Btw default experience nowadays is reflect metadata and no cache.

> From what I understand, the metadata cache is only generated through the CLI (and never from regular operations of the main application). 

Why do you think so? Caching works the very same way, regarldess how you initialize the ORM (CLI is a way to initialize it, under the hood it just calls `MikroORM.init` when it needs to read the metadata or use db connection). Cache gets generated when you first run the app. Production cache gets generated on the first run with production settings (so via `node`, not via `ts-node` or similar) - that is exactly what the generate cache CLI command do.

> I don't really want to load a different provider to read metadata, as I already have it.

I don't understand this.

> I could imagine providing a StrictFileCacheAdapter that's almost identical to current FileCacheAdapter, but throws on validation failures. In general, the adapter should be enriched with (debug) logging to surface any failures or mismatches.

So for you it is better DX to throw instead of regenerating if we see invalid cache? Why?

**Reply by oliversalzburg** [reply]
Thanks for the detailed response. As often, it clears up a lot of confusion on my part. It seems like I had a few false assumptions in regards to how the cache is supposed to work, due to misconfigurations on my behalf.

I was under the impression that, when using the `TsMorphMetadataProvider`, the cache should always be built prior to deployment, to avoid running the costly operations associated with it. In that case, I felt it would be in the developers best interest if the application would `throw` when deployed, rather than run the costly operation the developer was trying to avoid by deploying the cache in the first place.

It made sense to me to have the `TsMorphMetadataProvider` loaded only in the short-lived CLI and generate the cache out of the main application process. So that you can rely purely on the cache during deployment.

I can now see that that is just one way to use MikroORM in that configuration. Due to my misconfigurations, I was under the assumptions that it's the only way and felt the default cache adapter behavior was confusing.

**Reply by B4nan** [reply]
You might be interested in this thread too: https://github.com/mikro-orm/mikro-orm/discussions/2103

There is definitely room for improvement, I am not opposed to changing how things work, and especially supporting the use case from that linked discussion.

Also we might throw as you suggested, but only in production (based on `NODE_ENV`). It would have to be configurable for sure.

**Reply by oliversalzburg** [reply]
Yeah, that thread is very much in line with what I thought was the de-facto way to use the `TsMorphMetadataProvider`.

The fact that I'm not generating the DB schema from MikroORM really gives me a lot of chances to run into weird issues. You can get a setup with reading and writing data perfectly without having properly defined column types on a lot of your entity schemas.

So sometimes I wish for strictness or validation in areas that are probably not relevant to the normal user who will generate the schema from Mikro and see issues much earlier.

#2103 still feels like an overall improvement to me for regular users as well. Generating the metadata during production deployments should be avoided IMO. I'm currently achieving this with my own cache adapter.
It does feel like the best approach to deal with the metadata, as the ts-morph approach is the most data-rich and requires the least additional information in decorators. The only downside is the resource usage, but this can be worked around with the cache.

I feel like this approach should be favored over reflect-metadata, but throughout the time I used MikroORM, it seemed like the ts-morph approach was the most experimental and should be avoided. Not sure why I got this impression, but wanted to share regardless :)
